File: pagamento-ricorrente/python/pagamento_successivo.py
                                            
# -*- coding: utf-8 -*-

# Pagamento ricorrente - Pagamento successivo - Chiamata sincrona

import sys
if sys.version_info >= (3,):
    from urllib.parse import urlencode
else:
    from urllib import urlencode
import hashlib
import datetime
import time
import requests

requestUrl = "https://int-ecommerce.nexi.it/ecomm/api/recurring/pagamentoRicorrente";

# Alias e chiave segreta
APIKEY = "<ALIAS>" # Sostituire con il valore fornito da Nexi
CHIAVESEGRETA = "<CHIAVE SEGRETA PER CALCOLO MAC>" # Sostituire con il valore fornito da Nexi

# Parametri della richiesta
numContratto = "TESTPS_" + datetime.datetime.today().strftime('%Y%m%d%H%M%s')

codTrans = "TESTPS_" + datetime.datetime.today().strftime('%Y%m%d%H%M%s')
importo = "5000"
divisa = "978"
scadenza = '202012'
timeStamp = (int(time.time())) * 1000

# Calcolo MAC
mac_str = 'apiKey=' + APIKEY + \
    'numeroContratto=' + numContratto + \
    'codiceTransazione=' + codTrans + \
    'importo=' + importo + \
    "divisa=" + divisa + \
    "scadenza=" + scadenza + \
    "timeStamp=" + str(timeStamp) + \
     CHIAVESEGRETA
mac =  hashlib.sha1(mac_str.encode('utf8')).hexdigest()

requestParams = {
    'apiKey': APIKEY,
    'numeroContratto': numContratto,
    'codiceTransazione': codTrans,
    'importo': importo,
    'divisa': divisa,
    'scadenza': scadenza,
    'timeStamp': str(timeStamp),
    'mac': mac
}
import json
print(json.dumps(requestParams, sort_keys=True, indent=4, separators=(',', ': ')))

response  = requests.post(requestUrl,json=requestParams,headers={'Content-Type':'application/json'})
try:
    response_data = response.json()

    if response_data['esito'] == "OK": # Transazione andata a buon fine
        # calcolo MAC con i parametri di ritorno
        macResponse = 'esito=' + response_data['esito'] + 'idOperazione=' + response_data['idOperazione'] + 'timeStamp=' + response_data['timeStamp'] + CHIAVESEGRETA
        macCalculated =  hashlib.sha1(macResponse.encode('utf8')).hexdigest()

        if macCalculated != response_data['mac']:
            raise ValueError('Errore MAC: ' + macCalculated + ' non corrisponde a ' + response_data['mac'])

        print('La transazione ' + codTrans + " è avvenuta con successo; codice autorizzazione: " + response_data['codiceAutorizzazione'])
    else: # Transazione rifiutata
        print('La transazione ' + codTrans + " è stata rifiutata; descrizione errore: " + response_data['errore']['messaggio'])

except Exception as e:
    print(response)
    print(response.content)
                    
                

File: pagamento-ricorrente/java/pagamento_successivo.java
                                            
// Pagamento ricorrente - Pagamento successivo - Chiamata sincrona

import org.apache.http.HttpEntity;
import org.apache.http.HttpResponse;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.entity.StringEntity;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClientBuilder;
import org.apache.http.util.EntityUtils;
import org.json.JSONObject;

import java.security.MessageDigest;
import java.text.SimpleDateFormat;
import java.util.Date;

public class pagamentiSuccessvi-ChiamataSincrona {

    public static void main(String[] a) throws Exception {

        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyyMMddHHmmss");
        Date date = new Date();

        // URL + URI
        String requestUrl = "https://int-ecommerce.nexi.it/" + "ecomm/api/recurring/pagamentoRicorrente";

        // PARAMETRI PER CALCOLO MAC
        String apiKey = "<ALIAS>"; // Alias fornito da Nexi
        String numeroContratto = "3dsa3asc34a"; // Numero del contratto
        String codTrans = "23ds45s"; // Codice Transazione da eseguire
        String importo = "5000"; // 5000 = 50,00 Euro (cifre in centesimi)
        String divisa = "EUR"; // EUR = 978
        String scadenza = "203012"; // Scadenza aaaamm
        String codiceGruppo = "<GRUPPO>"; // Codice Gruppo fornito da Nexi
        String timeStamp = "" + System.currentTimeMillis();
        String chiaveSegreta = "<CHIAVE SEGRETA PER CALCOLO MAC>"; // Chiave segreta fornita da Nexi

        // CONCATENAMENTO STRINGA (MAC)
        String stringaMac = "apiKey=" + apiKey
                + "numeroContratto=" + numeroContratto
                + "codiceTransazione=" + codTrans
                + "importo=" + importo
                + "divisa=" + divisa
                + "scadenza=" + scadenza
                + "timeStamp=" + timeStamp
                + chiaveSegreta;

		
        // CALCOLO MAC
        String macCalculated = hashMac(stringaMac);

		// PARAMETRI DI INVIO	
        JSONObject json = new JSONObject();
        json.put("apiKey", apiKey);
        json.put("numeroContratto", numeroContratto);
        json.put("codiceTransazione", codTrans);
        json.put("importo", importo);
        json.put("divisa", divisa);
        json.put("scadenza", scadenza);
        json.put("timeStamp", timeStamp);
        json.put("mac", macCalculated);
        

        // CHIAMATA API
        CloseableHttpClient httpClient = HttpClientBuilder.create().build();
        HttpPost request = new HttpPost(requestUrl);
        StringEntity params = new StringEntity(json.toString());
        request.addHeader("content-type", "application/json");
        request.setEntity(params);
        HttpResponse response = httpClient.execute(request);
        HttpEntity entity = response.getEntity();
        String responseString = EntityUtils.toString(entity, "UTF-8");

        // PARAMETRI DI RITORNO
        JSONObject responseObj = new JSONObject(responseString);
    }

    public static String hashMac(String stringaMac) throws Exception {
        MessageDigest digest = MessageDigest.getInstance("SHA-1");
        byte[] in = digest.digest(stringaMac.getBytes("UTF-8"));

        final StringBuilder builder = new StringBuilder();
        for (byte b : in) {
            builder.append(String.format("%02x", b));
        }
        return builder.toString();
    }
}
                    
                